### ISSUE TITLE
Error integrating with fastMNN when specifying merge order

### ISSUE THREAD

**Issue by dkeitley** [LTLA/batchelor#30]
Hi,

I'm trying to perform batch correction using fastMNN with a hierarchical merge order which integrates within each developmental stage before merging across stages. I've run the following below but I'm running into a subscript out of bounds error. 

```
sce <- loadWagner2018()

gene_var <- modelGeneVar(sce)
hvgs <- getTopHVGs(gene_var,n=5000)
meta <- colData(sce)

order_df = meta[!duplicated(sce$library_id), c("stage", "library_id")]
order_df$ncells = sapply(order_df$library_id, function(x) sum(meta$library_id == x))
order_df$stage = factor(order_df$stage, 
                        levels = rev(c("24hpf", 
                                       "18hpf", 
                                       "14hpf", 
                                       "10hpf", 
                                       "8hpf", 
                                       "6hpf", 
                                       "4hpf")))
order_df$library_id <- as.factor(order_df$library_id)
order_df = order_df[order(order_df$stage, order_df$ncells, decreasing = TRUE),]

merge_order <- lapply(split(order_df,order_df$stage), function(x) list(x$library_id))
names(merge_order)<- NULL

order_df$stage = as.character(order_df$stage)


out <- fastMNN(sce,batch=sce$library_id,subset.row = hvgs,
               merge.order = rev(merge_order))
```

Error message:
`Error: subscript contains out-of-bounds indices`

After a bit of digging, it looks like the error is thrown while trying to reorder the MNN output at [line 387](https://github.com/LTLA/batchelor/blob/master/R/fastMNN.R#L387). 
```
# Reordering by the input order.        
    d.reo <- divided$reorder
    output <- output[d.reo,,drop=FALSE]
```

I was wondering if you could suggest what I might be doing wrong...? 

The dataset has 63,530 cells and so `d.reo` is a permutation of these, however, the dimensions of `output` seems to be `(29106, 2)` which I was a bit confused by. 

The merge.order list specified contains a nested list of the library IDs:
```
[[1]]
[[1]][[1]]
 [1] DEW106 DEW108 DEW105 DEW103 DEW102 DEW107 DEW109 DEW164 DEW101 DEW166 DEW104 DEW163 DEW110
[14] DEW165 DEW055 DEW162 DEW057 DEW056 DEW053 DEW054 DEW052 DEW168 DEW169 DEW021 DEW159 DEW158
[27] DEW167 DEW160 DEW161
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[2]]
[[2]][[1]]
[1] DEW003 DEW038 DEW039 DEW041 DEW040 DEW012 DEW001
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[3]]
[[3]][[1]]
[1] DEW035 DEW036 DEW037 DEW011
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[4]]
[[4]][[1]]
[1] DEW033 DEW034 DEW010 DEW032
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[5]]
[[5]][[1]]
[1] DEW048 DEW047 DEW049 DEW046
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[6]]
[[6]][[1]]
[1] DEW043 DEW044 DEW045 DEW042
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169


[[7]]
[[7]][[1]]
[1] DEW050 DEW051
54 Levels: DEW001 DEW003 DEW010 DEW011 DEW012 DEW021 DEW032 DEW033 DEW034 DEW035 ... DEW169

```

Any hits would be appreciated!

Best,

Dan 

___

Session info:
```
R version 4.0.4 (2021-02-15)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] batchelor_1.6.2             SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
 [4] Biobase_2.50.0              GenomicRanges_1.42.0        GenomeInfoDb_1.26.4        
 [7] BiocNeighbors_1.8.2         BiocParallel_1.24.1         DelayedArray_0.16.2        
[10] IRanges_2.24.1              S4Vectors_0.28.1            MatrixGenerics_1.2.1       
[13] matrixStats_0.58.0          BiocGenerics_0.36.0         Matrix_1.3-2               
[16] BiocSingular_1.6.0         

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6                BiocManager_1.30.10       compiler_4.0.4           
 [4] bluster_1.0.0             XVector_0.30.0            bitops_1.0-6             
 [7] tools_4.0.4               DelayedMatrixStats_1.12.3 zlibbioc_1.36.0          
[10] digest_0.6.27             statmod_1.4.35            evaluate_0.14            
[13] lattice_0.20-41           rlang_0.4.10              pkgconfig_2.0.3          
[16] igraph_1.2.6              yaml_2.2.1                xfun_0.21                
[19] GenomeInfoDbData_1.2.4    knitr_1.31                locfit_1.5-9.4           
[22] grid_4.0.4                scuttle_1.0.4             rmarkdown_2.7            
[25] limma_3.46.0              irlba_2.3.3               magrittr_2.0.1           
[28] edgeR_3.32.1              htmltools_0.5.1.1         sparseMatrixStats_1.2.1  
[31] beachmat_2.6.4            rsvd_1.0.3                dqrng_0.2.1              
[34] ResidualMatrix_1.0.0      RCurl_1.98-1.2            scran_1.18.5   

```

**Comment by LTLA** [comment]
Looks like `merge.order` doesn't like `factor`s. I've fixed this in 1.7.13 (also on `master`) but in the meantime, you can get around this by coercing `library_id` into a character vector before constructing `merge.order`.

**Comment by dkeitley** [comment]
Yep, that seems to have fixed it. Thanks a lot!
